README.md
## Build libvpx for Android
####
<table>
<thead>
<tr><td>library</td><td>version</td><td>platform support</td><td>arch support</td></tr>
</thead>
<tr><td>libvpx</td><td>1.10.0</td><td>android</td><td>armeabi-v7a arm64-v8a x86 x86_64</td></tr>
</table>

### Build For Android
- Follow the instructions below to build libvpx for android
- aTalk v2.6.1 release is compatible with libvpx-1.10.0 (use android-ndk-r18b) <br/>
- aTalk v1.8.2 release is compatible with libvpx-1.8.0 (must use android-ndk-r17c or lower) <br/>
- aTalk v2.3.2 release uses libvpx-1.8.2 (work with android-ndk-r17c or android-ndk-r18b)<br/>
  Following problem has been fixed with inclusion of configure option --disable-avx2<br/>
  see <https://bugs.chromium.org/p/webm/issues/detail?id=1623#c1><br/>
  i.e.: The compiled libjnvpx.so for aTalk has a problem when exec on x86_64 android platform (libvpx asm source errors):<br/>
  org.atalk.android A/libc: Fatal signal 31 (SIGSYS), code 1 in tid 5833 (Loop thread: ne), pid 4781 (g.atalk.android)
- For armeabi-v7a build, need to add --disable-neon-asm for libvpx v1.8.2, otherwise:<br/>
  --clang70: error: linker command failed with exit code 1 (use -v to see invocation)<br/>
  --./lib/crtbegin_dynamic.o: crtbegin.c:function _start_main: error: undefined reference to 'main'<br/>
  --make\[1]: *** \[vpx_dsp/arm/intrapred_neon_asm.asm.S.o] Error 1<br/>
  --make\[1]: *** \[vpx_dsp/arm/vpx_convolve_copy_neon_asm.asm.S.o] Error 1<br/>
- When you first exec build-libvpx4android.sh, it applies the required patches to libvpx<br/>
  Note: the patches defined in libvpx_patch.sh is for libvpx-1.8.0+, libvpx-1.7.0 and libvpx-1.6.1+<br/>
  
The ./build-libvpx4android.sh script builds the static libvpx.a for the various architectures as defined in ./_settings.sh<br/>
i.e. ABIS=("armeabi-v7a" "arm64-v8a" "x86" "x86_64")<br/>
All the built libvpx.a and *.h are installed in the ./android/&lt;ABI>/lib and include respectively

**Android libvpx build instructions:**
```
## Use Android NDK: android-ndk-r18b (libvpx v1.8.2)
export ANDROID_NDK=/opt/android/android-ndk-r18b

## setup the required libvpx; default "libvpx-1.10.0" or change LIB_GIT in ./init_libvpx.sh
Optional as next command will load the source if not found or incorrect version, and applied the patches if any
./init_libvpx.sh
 
## use one of the following to build libvpx i.e.
a. for all the ABI's defined in _settings.sh
./build-libvpx4android.sh

b. for a specific <ABI>
./build-libvpx4android.sh <ABI>

The build installs the static libs and includes into the android project jni directory
 i.e. aTalk/jni/vpx/android/\<ABI>
```

Copy the static libs and includes in `./output/android/<ABI>` directories to the android project
jni directory i.e. aTalk/jni/vpx/android/\<ABI>

Note: 20210517: For all future update, please refer to atalk-android static_library_build/libvpx scripts

```
Note:
## Standalone toolchains work for ABIS=("arm64-v8a" "x86" "x86_64")
ABIS "armeabi-v7a" has errors for libvpx-1.8.0 i.e.
clang50: error: unsupported option '--defsym'
clang50: error: no such file or directory: 'ARCHITECTURE=7'

The above problem is fixed in libvpx-1.8.2 (standalone toolChanins only: sdk-path option is removed)
```

All information given below is for reference only. See aTalk/jni for its implementation.

##### ============================================
#### Android.mk makefile - for other project
Add libvpx include path to `jni/Android.mk`. 

```
#Android.mk

include $(CLEAR_VARS)
LOCAL_MODULE := libvpx
LOCAL_SRC_FILES := Your libVPX Library Path/$(TARGET_ARCH_ABI)/libvpx.a
include $(PREBUILT_STATIC_LIBRARY)

LOCAL_C_INCLUDES := $(LOCAL_PATH)/Your libvpx Include Path/libvpx
LOCAL_STATIC_LIBRARIES := libvpx
LOCAL_LDLIBS := -lz
	
```

License
-------

    ffmpeg, android static library for aTalk VoIP and Instant Messaging client
    
    Copyright 2016 Eng Chong Meng
        
    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at
    
       http://www.apache.org/licenses/LICENSE-2.0
    
    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.






Changelog
# Entries are sorted chronologically from oldest to youngest within each release,
# releases are sorted from youngest to oldest.

#=============================================================
version 1.1.7 (2021/07/19):
- Update scripts for v1.10.0
- Update init_libvpx.sh to have smart libvpx fetching
- Update build-vpx4android.sh to only make clean on condition to speed delta build 

#=============================================================
version 1.1.6 (2021/04/23):
- Build using libvpx v1.10.0
- All other settings remain as with v1.8.2 without any changes

#=============================================================
version 1.1.5 (2020/07/15):
- Clean up scripts for libvpx v1.8.2 build

#=============================================================
version 1.1.4 (2019/05/24):
- Clean up scripts and use only single script execution to build native library

#=============================================================
version 1.1.3 (2019/04/23):
- Use option "--enable-runtime-cpu-detect" to fix x86_64 VP8 encoder problem
  see https://bugs.chromium.org/p/webm/issues/detail?id=1623#c1\
- cleanup scrips to allow build with standalone toolchains i.e.
  ## Standalone toolchains work for ABIS=("arm64-v8a" "x86" "x86_64")
  ABIS "armeabi-v7a" has errors i.e.
  clang50: error: unsupported option '--defsym'
  clang50: error: no such file or directory: 'ARCHITECTURE=7'

#=============================================================
version 1.1.2 (2019/04/20):
- Cleanup all build scripts for libvpx v1.8.0
- use libvpx v1.8.0 for aTalk
- add option --disable-avx2 to fix x86_64 VP8 encoder problem
  see https://bugs.chromium.org/p/webm/issues/detail?id=1623#c1

#=============================================================
version 1.1.1 (2019/04/13):
- Cleanup all build scripts and README

#=============================================================
version 1.1.0 (2019/03/21):
- Cleanup all build scripts, settings and README
- Add patches for libvpx 1.7.0 and 1.6.1+
- Patches fix build for arm64-v8a (incorrect gcc/g++)
- Patches fix x86 and x86-64 (relocation R_386_GOTOFF against preemptible symbol vpx_rv etc)

#=============================================================
version 1.0.0 (2018/03/20):
- initial release


_shared.sh
#!/bin/bash

# export ANDROID_NDK="/opt/android/android-ndk-r15c" - last working is r15c without errors
# r16b => Unable to invoke compiler: /opt/android/android-ndk-r16b/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/bin/arm-linux-androideabi-gcc but build?

if [ "$ANDROID_NDK" = "" ]; then
	echo "You need to set ANDROID_NDK environment variable, exiting"
	echo "Use: export ANDROID_NDK=/your/path/to/android-ndk"
	exit 1
fi

# Never mix two api level to build static library for use on the same apk.
# Set to API:15 for aTalk minimun support for platform API-15
# Does not build 64-bit arch if ANDROID_API is less than 21 - the minimum supported API level for 64-bit.
ANDROID_API=21
NDK_ABI_VERSION=4.9

# Do not change naming convention of the ABIS; see:
# https://developer.android.com/ndk/guides/abis.html#Native code in app packages
# Android recomended architecture support; others are deprecated
# ABIS=("armeabi-v7a" "arm64-v8a" "x86" "x86_64")
ABIS=("armeabi" "armeabi-v7a" "arm64-v8a" "x86" "x86_64" "mips" "mips64")

BASEDIR=`pwd`
NDK=${ANDROID_NDK}
HOST_NUM_CORES=$(nproc)


# Do not modify any of the NDK_ARCH, CPU and -march unless you are sure.
# The settings are used by <ARCH>-linux-android-gcc and submodule configure
# https://en.wikipedia.org/wiki/List_of_ARM_microarchitectures
# $NDK/toolchains/llvm/prebuilt/...../includellvm/ARMTargetParser.def etc
# ARCH - should be one from $ANDROID_NDK/platforms/android-$API/arch-* [arm / arm64 / mips / mips64 / x86 / x86_64]"

configure() {
  ABI=$1;
  TOOLCHAIN_PREFIX=${BASEDIR}/${ABI}-android-toolchain

  case $ABI in
    armeabi)
      # Deprecated in r16. Will be removed in r17
      NDK_ARCH="arm"
      export ARCH_LINK=""
      export NDK_ABIARCH="arm-linux-androideabi"
      ARCH_FLAGS="-march=armv5 -marm -finline-limit=64"
      ASFLAGS=""
    ;;
    armeabi-v7a)
      NDK_ARCH="arm"
      export ARCH_LINK="-march=armv7-a -Wl,--fix-cortex-a8"
      export NDK_ABIARCH="arm-linux-androideabi"
      ARCH_FLAGS="-march=armv7-a -mfloat-abi=softfp -mfpu=vfpv3-d16 -mthumb -mfpu=neon"
      ASFLAGS=""
    ;;
    arm64-v8a)
      # Valid cpu = armv8-a cortex-a35, cortex-a53, cortec-a57 etc. but -march=armv8-a is required
      # x264 build has own undefined references e.g. x264_8_pixel_sad_16x16_neon - show up when build ffmpeg 
      NDK_ARCH="arm64"
      export ARCH_LINK=""
      export NDK_ABIARCH="aarch64-linux-android"
      ARCH_FLAGS="-march=armv8-a -mfloat-abi=softfp -mfpu=vfpv3-d16 -mfpu=neon"
      ASFLAGS=""
    ;;
    x86)
      NDK_ARCH="x86"
      export ARCH_LINK=""
      export NDK_ABIARCH="i686-linux-android"
      ARCH_FLAGS="-O2 -march=i686 -mtune=intel -msse3 -mfpmath=sse -m32"
      ASFLAGS="-D__ANDROID__"
    ;;
    x86_64)
      NDK_ARCH="x86_64"
      export ARCH_LINK=""
      export NDK_ABIARCH="x86_64-linux-android"
      ARCH_FLAGS="-O2 -march=x86-64 -mtune=intel -msse4.2 -mpopcnt -m64"
      ASFLAGS="-D__ANDROID__"
    ;;
    mips)
      NDK_ARCH="mips"
      export ARCH_LINK=""
      export NDK_ABIARCH="mipsel-linux-android"
      ARCH_FLAGS=""
      ASFLAGS=""
    ;;
    mips64)
      NDK_ARCH="mips64"
      export ARCH_LINK=""
      export NDK_ABIARCH="mips64el-linux-android"
      ARCH_FLAGS=""
      ASFLAGS=""
    ;;
  esac

  # cmeng: must ensure AS JNI uses the same STL library or "system"
  [ -d ${TOOLCHAIN_PREFIX} ] || python $NDK/build/tools/make_standalone_toolchain.py \
     --arch ${NDK_ARCH} \
     --api ${ANDROID_API} \
     --stl libc++ \
     --install-dir=${TOOLCHAIN_PREFIX}

  SYSROOT=${TOOLCHAIN_PREFIX}/sysroot
  PREFIX=${BASEDIR}/output/android/${ABI}

  # export TOOLCHAIN_PATH=${TOOLCHAIN_PREFIX}/bin
  export PATH=$TOOLCHAIN_PREFIX/bin:$PATH
  export CROSS_PREFIX=${TOOLCHAIN_PREFIX}/bin/${NDK_ABIARCH}-
  export CC=${CROSS_PREFIX}clang
  export CXX=${CROSS_PREFIX}clang++
  export LINK=${CXX}
  export LD=${CROSS_PREFIX}ld
  export AR=${CROSS_PREFIX}ar
  export RANLIB=${CROSS_PREFIX}ranlib
  export STRIP=${CROSS_PREFIX}strip
  export LIBS=${LIBS:-""}
  export CFLAGS="${ARCH_FLAGS} -fpic -ffunction-sections -funwind-tables -fstack-protector -fno-strict-aliasing"
  export CXXFLAGS="${CFLAGS} -std=c++11 -frtti -fexceptions"
  export ASFLAGS="${ASFLAGS}"
  export LDFLAGS="${ARCH_LINK}"

  echo "**********************************************"
  echo "use ANDROID_API=${ANDROID_API}"
  echo "use NDK=${NDK}"
  echo "export ABI=${ABI}"
  echo "export CROSS_PREFIX=${CROSS_PREFIX}"
  # echo "export SYSROOT=${SYSROOT}"
  echo "export CC=${CC}"
  echo "export CXX=${CXX}"
  echo "export LINK=${LINK}"
  echo "export LD=${LD}"
  echo "export AR=${AR}"
  echo "export RANLIB=${RANLIB}"
  echo "export STRIP=${STRIP}"
  # echo "export CPPFLAGS=${CPPFLAGS}"
  echo "export CFLAGS=${CFLAGS}"
  echo "export CXXFLAGS=${CXXFLAGS}"
  echo "export LDFLAGS=${LDFLAGS}"
  echo "export LIBS=${LIBS}"
  echo "export ASFLAGS=${ASFLAGS}"
  echo "**********************************************"
}


libvpx_patch.sh
#!/bin/bash
# set -x
# Applying required patches for libvpx

if [[ $# -eq 1 ]]; then
  LIB_VPX=$1
else
  LIB_VPX="libvpx"
fi

if [[ -f "${LIB_VPX}/build/make/version.sh" ]]; then
  version=`"${LIB_VPX}/build/make/version.sh" --bare "${LIB_VPX}"`
else
  version='v1.10.0'
fi

echo -e "\n*** Applying patches for: ${LIB_VPX} (${version}) ***"

# ===============================
# libvpx patches for version 1.8.0, 1.7.0 and 1.6.1+
# None is applicable and are skipped for the libvpx version 1.10.0
# ===============================
if [[ (${version} < v1.10.0 ) ]]; then
  echo -e "\n*** Applying patches for: ${LIB_VPX} (${version}) ***"

  if [[ ! (v1.8.2 > ${version}) ]]; then
    patch  -p0 -N --dry-run --silent -f ./${LIB_VPX}/build/make/configure.sh < ./patches/10.libvpx_configure.sh.patch 1>/dev/null
    if [[ $? -eq 0 ]]; then
      patch -p0 -f ./${LIB_VPX}/build/make/configure.sh < ./patches/10.libvpx_configure.sh.patch
    fi
  fi

  # v1.8.0 does not have filter_x86.c
  if [[ "${version}" == v1.7.0 ]] || [[ "${version}" == v1.6.1 ]]; then
    patch  -p0 -N --dry-run --silent -f ./${LIB_VPX}/vp8/common/x86/filter_x86.c < ./patches/11.libvpx_filter_x86.c.patch 1>/dev/null
    if [[ $? -eq 0 ]]; then
      patch -p0 -f ./${LIB_VPX}/vp8/common/x86/filter_x86.c < ./patches/11.libvpx_filter_x86.c.patch
    fi

    patch  -p0 -N --dry-run --silent -f ./${LIB_VPX}/vpx_dsp/deblock.c < ./patches/12.libvpx_deblock.c.patch 1>/dev/null
    if [[ $? -eq 0 ]]; then
      patch -p0 -f ./${LIB_VPX}/vpx_dsp/deblock.c < ./patches/12.libvpx_deblock.c.patch
    fi

    patch  -p0 -N --dry-run --silent -f ./${LIB_VPX}/vpx_ports/mem.h < ./patches/13.libvpx_mem.h.patch 1>/dev/null
    if [[ $? -eq 0 ]]; then
      patch -p0 -f ./${LIB_VPX}/vpx_ports/mem.h < ./patches/13.libvpx_mem.h.patch
    fi
  fi
fi

# ===============================
# Patches for libvpx version 1.10.0
# v1.10.0 need below patch for vp9 encode to work properly; master copy has been fixed
# ===============================

if [[ "${version}" == v1.10.0 ]]; then
  patch  -p0 -N --dry-run --silent -f ./${LIB_VPX}/vpx/vpx_encoder.h < ./patches/10.vpx_encoder_h.patch 1>/dev/null
  if [[ $? -eq 0 ]]; then
    patch -p0 -f ./${LIB_VPX}/vpx/vpx_encoder.h < ./patches/10.vpx_encoder_h.patch
  fi
fi

build-vpx4android.sh
#!/bin/bash
#
# Copyright 2016 Eng Chong Meng
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#     http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# (20210521) aTalk v2.6.1 uses libvpx-1.10.0, and the scripts are verified for this version only
#set -x
set -u
. _settings.sh

# aTalk v1.8.1 and below uses libvpx-master i.e. 1.6.1+ (10/13/2017)
# aTalk v1.8.2 uses libvpx-1.8.0
# aTalk v2.3.2 uses libvpx-1.8.2
# aTalk v2.6.1 uses libvpx-1.10.0

LIB_VPX="libvpx"
LIB_GIT=v1.10.0

## Both problems #1 & #2 below have been fixed by patches from: https://android.googlesource.com/platform/external/libvpx/+/ca30a60d2d6fbab4ac07c63bfbf7bbbd1fe6a583
## However the compiled libjnvpx.so has problem when exec on x86_64 android platform:
## i.e. org.atalk.android A/libc: Fatal signal 31 (SIGSYS), code 1 in tid 5833 (Loop thread: ne), pid 4781 (g.atalk.android)

# 1. libvpx >v1.6.1 has the following errors when build with aTalk; v1.6.1 failed with --enable-pic
# ./i686-linux-android/bin/ld: error: vpx/android/x86/lib/libvpx.a(deblock_sse2.asm.o): relocation R_386_GOTOFF against preemptible symbol vpx_rv cannot be used when making a shared object
# ./i686-linux-android/4.9.x/../../../../i686-linux-android/bin/ld: error: vpx/android/x86/lib/libvpx.a(subpixel_mmx.asm.o): relocation R_386_GOTOFF against preemptible symbol # vp8_bilinear_filters_x86_8 cannot be used when making a shared object
# ./i686-linux-android/bin/ld: error: vpx/android/x86/lib/libvpx.a(subpixel_mmx.asm.o): relocation R_386_GOTOFF against preemptible symbol vp8_bilinear_filters_x86_8 cannot be used when making a shared object
# ./i686-linux-android/bin/ld: error: vpx/android/x86/lib/libvpx.a(subpixel_sse2.asm.o): relocation R_386_GOTOFF against preemptible symbol vp8_bilinear_filters_x86_8 cannot be used when making a shared object
# ./i686-linux-android/bin/ld: error: vpx/android/x86/lib/libvpx.a(subpixel_sse2.asm.o): relocation R_386_GOTOFF against preemptible symbol vp8_bilinear_filters_x86_8 cannot be used when making a shared object

# 2. However libvpx-1.6.1 x86_64 has the same error
# ./x86_64-linux-android/bin/ld: error: vpx/android/x86_64/lib/libvpx.a(deblock_sse2.asm.o): requires dynamic R_X86_64_PC32 reloc against 'vpx_rv' which may overflow at runtime; recompile with -fPIC

# Auto fetch and unarchive libvpx from online repository
./init_libvpx.sh ${LIB_GIT}

# Applying required patches to libvpx-1.10.0, libvpx-1.8.x, libvpx-1.7.0 or libvpx-1.6.1+
./libvpx_patch.sh ${LIB_VPX}

if [[ -f "${LIB_VPX}/build/make/version.sh" ]]; then
  version=`"${LIB_VPX}/build/make/version.sh" --bare "${LIB_VPX}"`
fi

# Unarchive library, then configure and make for specified architectures
configure_make() {
  pushd "${LIB_VPX}" || exit
  ABI=$1;
  echo -e "\n** BUILD STARTED: ${LIB_VPX} (${version}) for ${ABI} **"

  configure "$@"
  case ${ABI} in
    # libvpx does not provide armv5 build option
    armeabi)
      TARGET="armv7-android-gcc --disable-neon --disable-neon-asm"
    ;;
    # need to add --disable-neon-asm for libvpx v1.8.2, otherwise:
    # clang70: error: linker command failed with exit code 1 (use -v to see invocation)
	# ./lib/crtbegin_dynamic.o:crtbegin.c:function _start_main: error: undefined reference to 'main'
    # make[1]: *** [vpx_dsp/arm/intrapred_neon_asm.asm.S.o] Error 1
  	# make[1]: *** [vpx_dsp/arm/vpx_convolve_copy_neon_asm.asm.S.o] Error 1
	armeabi-v7a)
      TARGET="armv7-android-gcc --enable-neon --disable-neon-asm"
    ;;
    arm64-v8a)
      TARGET="arm64-android-gcc"
    ;;
    x86)
      TARGET="x86-android-gcc"
    ;;
    x86_64)
      TARGET="x86_64-android-gcc"
    ;;
    mips)
      TARGET="mips32-linux-gcc"
    ;;
    mips64)
      TARGET="mips64-linux-gcc"
    ;;
  esac

  # --sdk-path=${TOOLCHAIN_PREFIX} must use ${NDK} actual path else cannot find CC for arm64-android-gcc
  # ==> (Unable to invoke compiler: /arm-linux-androideabi-gcc)
  # https://bugs.chromium.org/p/webm/issues/detail?id=1476
  
  # fixed by patch 10.libvpx_configure.sh.patch
  # https://github.com/google/ExoPlayer/issues/3520 (VP9 builds failure with android-ndk-r16 #3520)
  # https://github.com/android-ndk/ndk/issues/190#issuecomment-375164450 (unknown type name __uint128_t on ndk-build #190)
  # Has configure error with Target=arm64-android-gcc which uses incorrect cc i.e. arm-linux-androideabi-gcc;

  # ./asm/sigcontext.h:39:3: error: unknown type name '__uint128_t'
  # GCC has builtin support for the types __int128, unsigned __int128, __int128_t and __uint128_t. Use them to define your own types:
  # typedef __int128 int128_t;
  # typedef unsigned __int128 uint128_t;
  # Standalone toolchains fixed the problem?

  # need --as=yasm which is required by x86 and x86-64; cannot use define in _settings.sh which uses clang
  # see https://github.com/webmproject/libvpx

  # --sdk-path=${NDK} when specified - configure will use SDK toolchains and gcc/g++ as the default compiler/linker
  # must specified --extra-cflags and --libc if use --sdk-path
  # --sdk-path=${NDK} \
  # --extra-cflags="-isystem ${NDK}/sysroot/usr/include/${NDK_ABIARCH} -isystem ${NDK}/sysroot/usr/include" \
  # must specified -libc from standalone toolchains, libvpx configure.sh cannot get the right arch to use

  # SDK toolchains has error with ndk-r18b; however ndk-R17c and ndk-r16b are ok (gcc/g++)
  # SDK toolchains ndk-r18b is working with libvpx v1.10.0 & v1.8.2 without the sdk option

  # Standalone toolchains built has problem with ABIS="armeabi-v7a"
  # /tmp/vpx-conf-31901-2664.o(.ARM.exidx.text.main+0x0): error: undefined reference to '__aeabi_unwind_cpp_pr0'
  #
  # Cannot define option add_ldflags "-Wl,--fix-cortex-a8"
  # Standalone: arm-linux-androideabi-ld: -Wl,--fix-cortex-a8: unknown option

  # Fixed by: https://android.googlesource.com/platform/external/libvpx/+/ca30a60d2d6fbab4ac07c63bfbf7bbbd1fe6a583
  # libvpx has the following errors for x86 and x86_64 when build in aTalk app;
  # ./i686-linux-android/bin/ld: error: vpx/android/x86/lib/libvpx.a(deblock_sse2.asm.o): relocation R_386_GOTOFF against preemptible symbol vpx_rv cannot be used when making a shared object
  # ./i686-linux-android/4.9.x/../../../../i686-linux-android/bin/ld: error: vpx/android/x86/lib/libvpx.a(subpixel_mmx.asm.o): relocation R_386_GOTOFF against preemptible symbol # vp8_bilinear_filters_x86_8 cannot be used when making a shared object
  # ./i686-linux-android/bin/ld: error: vpx/android/x86/lib/libvpx.a(subpixel_mmx.asm.o): relocation R_386_GOTOFF against preemptible symbol vp8_bilinear_filters_x86_8 cannot be used when making a shared object
  # ./i686-linux-android/bin/ld: error: vpx/android/x86/lib/libvpx.a(subpixel_sse2.asm.o): relocation R_386_GOTOFF against preemptible symbol vp8_bilinear_filters_x86_8 cannot be used when making a shared object
  # ./i686-linux-android/bin/ld: error: vpx/android/x86/lib/libvpx.a(subpixel_sse2.asm.o): relocation R_386_GOTOFF against preemptible symbol vp8_bilinear_filters_x86_8 cannot be used when making a shared object
  # ./x86_64-linux-android/bin/ld: error: vpx/android/x86_64/lib/libvpx.a(deblock_sse2.asm.o): requires dynamic R_X86_64_PC32 reloc against 'vpx_rv' which may overflow at runtime; recompile with -fPIC

  # Need --disable-avx2 to fix x86_64 problem OR enable --enable-runtime-cpu-detect option
  # org.atalk.android A/libc: Fatal signal 4 (SIGILL), code 2 (ILL_ILLOPN), fault addr 0x77b2ac1757e6 in tid 20780 (Loop thread: ne), pid 20363 (g.atalk.android)
  # see https://bugs.chromium.org/p/webm/issues/detail?id=1623#c1
  # OR use option --enable-runtime-cpu-detect for x86/x86_64 ABIS platforms


  if [[ -f "config.log" ]]; then
      config_target="$(grep "Configuring for target" < "config.log" | sed "s/^.* '\(.*\)'$/\1/")"
      if [[ ${TARGET} != "${config_target}" ]]; then
        make clean
      fi
  else
    make clean
  fi

CPU_DETECT="--disable-runtime-cpu-detect"
if [[ $1 =~ x86.* ]]; then
   CPU_DETECT="--enable-runtime-cpu-detect"
fi

# When use --sdk-path option for libvpx v1.8.0; must use android-ndk-r17c or lower
# For libvpx v1.8.2: in order to use standalone toolchanis, must not specified --sdk-path (option removed)
#    --sdk-path=${NDK}

  ./configure \
    --extra-cflags="-isystem ${NDK_SYSROOT}/usr/include/${NDK_ABIARCH} -isystem ${NDK_SYSROOT}/usr/include" \
    --libc=${NDK_SYSROOT} \
    --prefix=${PREFIX} \
    --target=${TARGET} \
    ${CPU_DETECT} \
    --as=yasm \
    --enable-pic \
    --disable-docs \
    --enable-static \
    --enable-libyuv \
    --disable-examples \
    --disable-tools \
    --disable-debug \
    --disable-unit-tests \
    --enable-realtime-only \
    --enable-vp8 --enable-vp9 \
    --enable-vp9-postproc \
    --enable-vp9-highbitdepth \
    --disable-webm-io || exit 1

  make -j${HOST_NUM_CORES} install
  popd || true
}

for ((i=0; i < ${#ABIS[@]}; i++))
do
  if [[ $# -eq 0 ]] || [[ "$1" == "${ABIS[i]}" ]]; then
    # Do not build 64 bit arch if ANDROID_API is less than 21 which is
    # the minimum supported API level for 64 bit.
    [[ ${ANDROID_API} -lt 21 ]] && ( echo "${ABIS[i]}" | grep 64 > /dev/null ) && continue;
    configure_make "${ABIS[i]}"
    echo -e "** BUILD COMPLETED: ${LIB_VPX} for ${ABIS[i]} **\n\n"
  fi
done
